src/Inventory.zig

```diff
@@ -507,69 +510,167 @@ pub const Command = struct { // MARK: Command
 		},
 	};
 
-	const SyncOperation = struct { // MARK: SyncOperation
+	const SyncOperationType = enum(u8) {
+		create = 0,
+		delete = 1,
+		useDurability = 2,
+	};
+
+	const SyncOperation = union(SyncOperationType) { // MARK: SyncOperation
 		// Since the client doesn't know about all inventories, we can only use create(+amount)/delete(-amount) and use durability operations to apply the server side updates.
-		inv: InventoryAndSlot,
-		amount: i32,
-		item: ?Item,
-		durability: i32,
+		create: struct {
+			inv: InventoryAndSlot,
+			amount: u16,
+			item: ?Item
+		},
+		delete: struct {
+			inv: InventoryAndSlot,
+			amount: u16
+		},
+		useDurability: struct {
+			inv: InventoryAndSlot,
+			durability: u32
+		},
 
 		pub fn executeFromData(data: []const u8) !void {
-			std.debug.assert(data.len >= 16);
-			var self = SyncOperation {
-				.inv = try InventoryAndSlot.read(data[0..8], .client, null),
-				.amount = std.mem.readInt(i32, data[8..12], .big),
-				.durability = std.mem.readInt(i32, data[12..16], .big),
-				.item = null,
-			};
-			if(data.len > 16) {
-				const zon = ZonElement.parseFromString(main.stackAllocator, data[16..]);
-				defer zon.deinit(main.stackAllocator);
-				self.item = try Item.init(zon);
-			}
-			if(self.amount > 0) { // Create
-				if(self.item) |item| {
-					self.inv.ref().item = item;
-				} else if(self.inv.ref().item == null) {
-					return error.Invalid;
-				}
-				if(self.inv.ref().amount +| self.amount > self.inv.ref().item.?.stackSize()) {
-					return error.Invalid;
-				}
-				self.inv.ref().amount += @intCast(self.amount);
-			} else if(self.amount < 0) { // Delete
-				if(self.inv.ref().amount < -self.amount) {
-					return error.Invalid;
-				}
-				self.inv.ref().amount -= @intCast(-self.amount);
-				if(self.inv.ref().amount == 0) {
-					self.inv.ref().item = null;
+			std.debug.assert(data.len >= 1);
+
+			switch (try deserialize(data)) {
+				.create => |create| {
+					if (create.item) |item| {
+						create.inv.ref().item = item;
+					} else if (create.inv.ref().item == null) {
+						return error.Invalid;
+					}
+
+					if (create.inv.ref().amount +| create.amount > create.inv.ref().item.?.stackSize()) {
+						return error.Invalid;
+					}
+					create.inv.ref().amount += create.amount;
+					
+					create.inv.inv.update();
+				},
+				.delete => |delete| {
+					if (delete.inv.ref().amount < delete.amount) {
+						return error.Invalid;
+					}
+					delete.inv.ref().amount -= delete.amount;
+					if (delete.inv.ref().amount == 0) {
+						delete.inv.ref().item = null;
+					}
+					
+					delete.inv.inv.update();
+				},
+				.useDurability => |durability| {
+					durability.inv.ref().item.?.tool.durability = @intCast(durability.durability);
```

Comment:
Should be `-|=` like it was before.
Also instead of doing the intCast, could you not just use @TypeOf(.durability) directly in the SyncOperation?